grabbit/src/test/groovy/com/twcable/grabbit/jcr/ProtoPropertyDecoratorSpec.groovy

```diff
@@ -0,0 +1,210 @@
+package com.twcable.grabbit.jcr
+
+import com.google.protobuf.ByteString
+import com.twcable.grabbit.proto.NodeProtos.Property as PropertyProto
+import com.twcable.grabbit.proto.NodeProtos.Value as ProtoValue
+import com.twcable.grabbit.proto.NodeProtos.Values as ProtoValues
+import org.apache.jackrabbit.value.BinaryValue
+import org.apache.jackrabbit.value.DateValue
+import org.apache.jackrabbit.value.StringValue
+import spock.lang.Specification
+
+import javax.jcr.Node
+import javax.jcr.Property
+import javax.jcr.PropertyType
+import javax.jcr.Value
+import javax.jcr.ValueFormatException
+
+import static org.apache.jackrabbit.JcrConstants.JCR_MIXINTYPES
+import static org.apache.jackrabbit.JcrConstants.JCR_PRIMARYTYPE
+
+class ProtoPropertyDecoratorSpec extends Specification {
+
+    def "When a property is either jcr:primaryType or jcr:mixinTypes, decorator will refuse to write these nodes are regular properties"() {
+        given:
+        final mockNode = Mock(Node)
+
+        when:
+        PropertyProto someProperty = PropertyProto.newBuilder()
+                                        .setName(refusedType)
+                                        .setType(PropertyType.STRING)
+                                        .setValue(
+                                            ProtoValue.newBuilder()
+                                                .setStringValue("somevalue")
+                                                .build()
+                                        )
+                                        .build()
+
+        final propertyDecorator = new ProtoPropertyDecorator(someProperty)
+        propertyDecorator.writeToNode(mockNode)
+
+        then:
+        thrown(IllegalStateException)
+
+        where:
+        refusedType << [JCR_PRIMARYTYPE, JCR_MIXINTYPES]
+    }
+
+
+    def "A single-value property can be written correctly to a node that expects a single-value property"() {
+        given:
+        final mockNode = Mock(Node)
+
+        when:
+        PropertyProto someProperty = PropertyProto.newBuilder()
+                                            .setName("someproperty")
+                                            .setType(PropertyType.STRING)
+                                            .setValue(
+                                                ProtoValue.newBuilder()
+                                                    .setStringValue("somevalue")
+                                                    .build()
+                                            )
+                                            .build()
+
+        final propertyDecorator = new ProtoPropertyDecorator(someProperty)
+        propertyDecorator.writeToNode(mockNode)
+
+        then:
+        1 * mockNode.setProperty("someproperty", new StringValue("somevalue"), PropertyType.STRING)
```

Comment:
I take the point of "I doubt `duplicate` + `hardcoded` = `clarity`", but in this specific case I agree with Jeff -- it's clearer/cleaner than the alternatives.

To take the use of `where:`, the "data-driven" approach makes things **_more_** confusing when used gratuitously, such as when there's only one data-set. It would be much better to simply declare them as variables/constants at the start of the method since they will only have one value, and it's much more idiomatic and shorter than declaring the `where:` section.

Even taking the simpler "declare a variable" approach, that adds more to the "mental stack" of things to keep track of, as well as adding the need to "dereference" the variable when reading it. Neither is a big deal, but it does have a small cost.

Now consider why "magic numbers" (of which hard coded duplicated values of any kind fall into the same category) are a bad thing: 1) you want a semantic clue what the value means, and 2) it's useful to have a common reference when the same logical thing is used in multiple places through the codebase (both to aid in changes, but also just to follow coherence).

For # 1 in the situation, the context and the value itself makes it more obvious than a name would. And for # 2 it's all scoped within a single (short) method.

I agree that duplicated hard-coded values are a pretty bad code-smell, but like all "smells" it doesn't mean it's _ipso facto_ wrong. Sometimes it's worth considering it compared to its alternatives and thinking more about what makes the pattern smelly. In **this particular case** it doesn't run into the problems with magic numbers, and the alternatives add further indirection with no clear benefit.